group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -1731,15 +1735,19 @@ public void onNewMetadataImage(
         // Push an event for each coordinator.
         coordinators.keySet().forEach(tp -> {
             scheduleInternalOperation("UpdateImage(tp=" + tp + ", offset=" + newImage.offset() + ")", tp, () -> {
-                withContextOrThrow(tp, context -> {
-                    if (context.state == CoordinatorState.ACTIVE) {
+                CoordinatorContext context = coordinators.get(tp);
+                if (context != null && context.state == CoordinatorState.ACTIVE) {
```

Comment:
I noticed there are still some places we don't want the context to be null -- ie loading the partition. (And for all the run methods we want the coordinator to be active) Is the general rationale for this change that not having the conetext/active state is ok if we expect the coordinator moved? Is it possible we could try to load and the coordinator moves again? Just trying to understand the cases we accept null or non-active coordinators vs when we throw errors.

Reply:
Good question. Generally, we never want the context to be `null` in internal operations. The only case where it could be is when the context is loaded. Then, most of the time, the context must be active in order to alter it. The only exception is during the loading/unloading. I have changed all the internal operations to act appropriately and log too.

The read and write events keep using `withActiveContextOrThrow`. In those case, throwing an exception is fine as it will complete the event with it.

Reply:
Ok -- so I see we updated the PR to match this description. I will take another pass 👍

Reply:
As an aside, I noticed there is one place we don't lock the context when accessing it and that is when closing. Is this to ensure the close happens as quickly as possible/doesn't get blocked?

Reply:
Oh. Nice catch. This is a bug. Let me fix it.